### ISSUE TITLE
False positive: unused_declaration for type member captured in closure

### ISSUE THREAD

**Issue by AleMaMi** [realm/SwiftLint#3332]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

Analyzer rule `unused_declaration` triggers false positive for type member used only as capture in closure

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
$ swiftlint analyze
```

Output is based on internal production code, cannot publish. But example should be clear enough.

### Environment

* SwiftLint version (run `swiftlint version` to be sure)? 0.40.1
* Installation method used (Homebrew, CocoaPods, building from source, etc)? Homebrew
* Paste your configuration file:

```yml
included:
  - Sources
  - Tests
analyzer_rules:
  - unused_declaration
  - unused_import
```

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)? No
  If so, paste their relative paths and respective contents.
* Which Xcode version are you using (check `xcodebuild -version`)? Xcode 11.7
Build version 11E801a
* Do you have a sample that shows the issue? Run `echo "[string here]" | swiftlint lint --no-cache --use-stdin --enable-all-rules`
  to quickly test if your example is really demonstrating the issue. If your example is more
  complex, you can use `swiftlint lint --path [file here] --no-cache --enable-all-rules`.

```swift
// This triggers a violation:
public struct FalsePositive {

    let value: String  // false positive here

    public func perform() -> () -> Void {
        { [value] in print("\(value)") }
    }
}
```

**Comment by jpsim** [comment]
This is a bug, but one that'll be tricky to fix.

You can work around this by re-binding `value` in your closure's scope.

E.g.

```swift
public struct FalsePositive {

    let value: String  // false positive here

    public func perform() -> () -> Void {
        { [value = self.value] in print("\(value)") }
    }
}
```
